### PR TITLE
USB: first step at unifying legacy and composite

### PR DESCRIPTION
**Pull request by aurel32**
This patchset is the first step trying to solve issue #10503. It unifies the standard, class and vendor request handler buffer, so that the same code is used in both cases.

Note that it is based on PR #11027 and thus includes the corresponding commits. I'll rebase that PR once the other get accepted.

### CODE REVIEW

**Review by finikorg — COMMENTED** [review]
I like the idea, maybe some patches may be merged since they change the same code, but if this is difficult could be used as is. For the netusb size error you may look at `samples/net/http_server/overlay-netusb.conf`, maybe we need to disable something.

File: subsys/usb/class/usb_dfu.c [hunk 1]
```diff
@@ -56,13 +56,7 @@
...
 
 #define NUMOF_ALTERNATE_SETTINGS	2
-
-#ifdef CONFIG_USB_COMPOSITE_DEVICE
-#define USB_DFU_MAX_XFER_SIZE		(min(CONFIG_USB_COMPOSITE_BUFFER_SIZE, \
-					     CONFIG_USB_DFU_MAX_XFER_SIZE))
-#else
-#define USB_DFU_MAX_XFER_SIZE		CONFIG_USB_DFU_MAX_XFER_SIZE
-#endif
+#define USB_DFU_MAX_XFER_SIZE		CONFIG_USB_REQUEST_BUFFER_SIZE
```
[the remark this answers is not in the document]
Comment (aurel32) [reply subsys/usb/class/usb_dfu.c]: The first few patches are from PR #11027, this PR is based on top of it. I believe that we should fix DFU in composite mode independently of merging legacy and composite code, which might need more discussion.

**Comment by aurel32** [comment]
> I like the idea, maybe some patches may be merged since they change the same code, but if this is difficult could be used as is.

That's not a problem to merge patches before merging the PR. I believe it might be better to keep them separated for the review, so we can accept only some of them and drop the other depending on the discussion.

> For the netusb size error you may look at `samples/net/http_server/overlay-netusb.conf`, maybe we need to disable something.

The ROM region overflow has been affecting many PR recently, but the problem seems to have gone. I guess the a recheck might help.

**Comment by aurel32** [comment]
recheck

**Comment by pfalcon** [comment]
Needs rebase.

**Comment by carlescufi** [comment]
@aurel32 would it make sense to close this in favor of https://github.com/zephyrproject-rtos/zephyr/pull/12682?

**Comment by aurel32** [comment]
> @aurel32 would it make sense to close this in favor of #12682?

From the quick look I got at #12682, there is a significant overlap between the two PR. I'll check that more in details when I have more time and either close this PR or submit the missing parts.

**Comment by jfischer-phytec-iot** [comment]
@aurel32 can you please rebase.

**Comment by nashif** [comment]
closing due to overlap with #12682. Reopen if still relevant.

**Comment by finikorg** [comment]
@aurel32 @jfischer-phytec-iot @nashif cherry-picked and rebased (PR is shown above)

**Pull request closed without merging**